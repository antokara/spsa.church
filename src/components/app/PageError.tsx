import { Box, Card, CardContent, Grid, WithTheme } from '@material-ui/core';
import { PaperProps } from '@material-ui/core/Paper';
import { Warning } from '@material-ui/icons';
import * as React from 'react';
import { default as styled } from 'styled-components';

/**
 * a styled card component that is need to create and pass the className prop
 */
const ClassedCard: React.ComponentType<PaperProps> = styled(Card)`
  background: ${(p: PaperProps & WithTheme): string =>
    p.theme.palette.error.dark};
`;

/**
 * the error card which uses the className provided by the "ClassedCard" styled component
 */
const ErrorCard: React.ComponentType<PaperProps> = ({
  children,
  className
}: React.PropsWithChildren<PaperProps>): JSX.Element => (
  <ClassedCard classes={{ root: className }}>{children}</ClassedCard>
);

/**
 * @prop error the error to show
 */
type TProps = {
  error: Error;
};

/**
 * a page error indicator, that covers the whole page
 * with absolute left/top positioning
 */
const PageError: (props: TProps) => JSX.Element = ({
  error
}: TProps): JSX.Element => (
  <Box
    display="flex"
    justifyContent="center"
    alignItems="center"
    position="absolute"
    top={0}
    left={0}
    width="100%"
    height="100%"
    bgcolor="background.paper"
  >
    <ErrorCard>
      <CardContent>
        <Grid container={true} spacing={1}>
          <Grid item={true} xs={1}>
            <Warning />
          </Grid>
          <Grid item={true} xs={11}>
            {error.message}
          </Grid>
        </Grid>
      </CardContent>
    </ErrorCard>
  </Box>
);

export { PageError };
