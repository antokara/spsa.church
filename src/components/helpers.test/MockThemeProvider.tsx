import { MuiThemeProvider, Theme } from '@material-ui/core';
import * as React from 'react';
import { THEME } from 'src/constants/THEME';
import { ThemeProvider } from 'styled-components';

type TProps = React.PropsWithChildren<{
  theme?: Theme;
}>;

/**
 * a theme provider for both MaterialUI and Styled Components
 *
 * It can be used to test (styled) components
 * that need the theme prop easily.
 *
 * Unless a theme is specified, it will use the app's theme
 */
const MockThemeProvider: React.FunctionComponent<TProps> = ({
  children,
  theme = THEME
}: TProps): React.ReactElement<React.ReactNode> => (
  <MuiThemeProvider theme={theme}>
    <ThemeProvider theme={theme}>{children}</ThemeProvider>
  </MuiThemeProvider>
);

export { MockThemeProvider };
