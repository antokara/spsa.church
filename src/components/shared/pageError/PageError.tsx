import { default as Box } from '@material-ui/core/Box';
import { CardProps, default as Card } from '@material-ui/core/Card';
import { default as CardContent } from '@material-ui/core/CardContent';
import { default as Grid } from '@material-ui/core/Grid';
import { WithTheme } from '@material-ui/core/styles';
import { default as Warning } from '@material-ui/icons/Warning';
import * as React from 'react';
import { default as styled } from 'styled-components';

/**
 * a styled card component that is need to create and pass the className prop
 */
const ClassedCard: React.ComponentType<CardProps> = styled(Card)`
  background: ${(p: CardProps & WithTheme): string =>
    p.theme.palette.error.dark};
`;

/**
 * the error card which uses the className provided by the "ClassedCard" styled component
 */
const ErrorCard: React.ComponentType<CardProps> = ({
  children,
  className
}: React.PropsWithChildren<CardProps>): JSX.Element => (
  <ClassedCard classes={{ root: className }}>{children}</ClassedCard>
);

/**
 * @prop error the error to show
 */
type TProps = {
  error: Error;
};

/**
 * a page error indicator, that covers the whole page
 * with absolute left/top positioning
 */
const PageError: (props: TProps) => JSX.Element = ({
  error
}: TProps): JSX.Element => (
  <Box
    display="flex"
    justifyContent="center"
    alignItems="center"
    position="absolute"
    top={0}
    left={0}
    width="100%"
    height="100%"
    bgcolor="background.paper"
  >
    <ErrorCard>
      <CardContent>
        <Grid container={true} spacing={1}>
          <Grid item={true} xs={1}>
            <Warning data-testid="warning-icon" />
          </Grid>
          <Grid item={true} xs={11}>
            {error.message}
          </Grid>
        </Grid>
      </CardContent>
    </ErrorCard>
  </Box>
);

export { PageError };
