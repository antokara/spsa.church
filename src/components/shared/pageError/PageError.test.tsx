import { render, RenderResult } from '@testing-library/react';
import * as React from 'react';
import { MockThemeProvider } from 'src/components/helpers.test/MockThemeProvider';
import { PageError } from './PageError';

describe('PageError component', () => {
  let rr: RenderResult;
  let node: ChildNode | null;
  let error: Error;

  beforeEach(() => {
    error = new Error('test error');
    rr = render(
      <MockThemeProvider>
        <PageError error={error} />
      </MockThemeProvider>
    );
    node = rr.container.firstChild;
  });

  it('renders the component', () => {
    expect(node).toMatchSnapshot();
  });

  it('renders the warning icon', () => {
    expect(rr.queryByTestId('warning-icon')).toBeTruthy();
  });

  it('renders the error message', () => {
    expect(rr.queryByText(error.message)).toBeTruthy();
  });
});
