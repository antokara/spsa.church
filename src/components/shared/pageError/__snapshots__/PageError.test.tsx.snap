// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageError component renders the component 1`] = `
<div
  class="MuiBox-root MuiBox-root-2"
>
  <div
    class="MuiPaper-root MuiCard-root sc-AxjAm cqvqZZ MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            data-testid="warning-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
            />
          </svg>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-11"
        >
          test error
        </div>
      </div>
    </div>
  </div>
</div>
`;
